### PR TITLE
Problem:(cro-819) no C bindings for client libraries

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: add c-bindings

### CODE REVIEW

File: cro-clib/src/bindings.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
+/* automatically generated by rust-bindgen */
```
Comment (yihuang) [inline cro-clib/src/bindings.rs]: Maybe don't commit this file since it's generated on the fly?

File: cro-clib/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+use client_common::storage::MemoryStorage;
+#[derive(Default)]
+pub struct ApiInfo {
+    pub chain_id: String,
+}
+static mut G_DB: Option<MemoryStorage> = None;
+
+/// # Safety
+/// this function should not be called inside rust
+pub unsafe fn get_string(src: *const ::std::os::raw::c_char) -> String {
```
Comment (yihuang) [inline cro-clib/src/lib.rs]: Do we need to make this `pub`?

File: cro-clib/chain.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+/* initialize state */
+typedef struct {
+    int error;
+    Buf chain_id;
+    Buf server_url;
+    Buf storage_folder;
+} ApiContext;
+
+extern ApiContext initialize(const char* chain_id, const char* server, const char* storage);
```
Comment (yihuang) [inline cro-clib/chain.h]:
```suggestion
extern "C" ApiContext initialize(const char* chain_id, const char* server, const char* storage);
```

File: cro-clib/chain.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+extern ApiResult get_network_id();
+extern HDWallet create_hdwallet(const char* name, const char* passphrase); /* hd wallet */
```
Comment (yihuang) [inline cro-clib/chain.h]:
```suggestion
extern HDWallet* create_hdwallet(const char* name, const char* passphrase); /* hd wallet */
```

Comment (leejw51crypto) [reply cro-clib/chain.h]: it's intended, not to use heap

File: cro-clib/chain.h (hunk 3 above)
Comment (leejw51crypto) [reply cro-clib/chain.h]: this is auto-matically generated by bindgen, but i'll check

File: cro-clib/src/lib.rs (hunk 2 above)
Comment (leejw51crypto) [reply cro-clib/src/lib.rs]: for now, no, but later , if we use separate files, it's necessary.

File: cro-clib/src/bindings.rs (hunk 1 above)
Comment (leejw51crypto) [reply cro-clib/src/bindings.rs]: that's correct, i'll remove it.

File: cro-clib/chain.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#define MAX_STRING_LENGTH 512
+/* general byte buffer */
+typedef struct {
+    unsigned char buf[MAX_STRING_LENGTH];
+    int length;
+} Buf;
+
+/* general api result */
+typedef struct {
+    int error;
```
Comment (yihuang) [inline cro-clib/chain.h]: I suggest following Cardano's style (https://github.com/input-output-hk/rust-cardano/blob/master/cardano-c/cardano.h#L183) to handle the error.

Comment (leejw51crypto) [reply cro-clib/chain.h]: ok, i'll check, 
but using heap is difficult to use.

File: cro-clib/chain.h (hunk 4 above)
Comment (yihuang) [reply cro-clib/chain.h]: > it's intended, not to use heap

A `Buf` is 516 bytes, a single `HDWallet` is already 3096 bytes, it's more costly than using the heap.

File: cro-clib/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        error: 0,
+        value: get_buf("wallet"),
+    }
+}
+
+/// # Safety
+/// this function should not be called inside rust
+pub unsafe fn do_create_hdwallet(
+    name: *const ::std::os::raw::c_char,
+    passphrase: *const ::std::os::raw::c_char,
```
Comment (tomtau) [inline cro-clib/src/lib.rs]: `use std::os::raw::c_char`?

File: cro-clib/chain.h (hunk 5 above)
Comment (tomtau) [reply cro-clib/chain.h]: I suggested that yesterday

File: cro-clib/src/lib.rs (hunk 6 above)
Comment (leejw51crypto) [reply cro-clib/src/lib.rs]: c-bindgen doesn't process struct,
output is the same.
i'll change to use heap.

**Comment by devashishdxt** [comment]
> * `bindgen` is for interfacing to an existing C library (C->Rust)  -- for generating C bindings (Rust->C, which is what we want), there's `cbindgen`

I Agree. We should generate C bindings from Rust code.

**Pull request closed without merging**